// webpack-config/dist/index.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.FILENAME = exports.patternCache = exports.cache = exports.environment = exports.Config = undefined;

var _Config = require('./Config');

var _Config2 = _interopRequireDefault(_Config);

var _ConfigEnvironment = require('./ConfigEnvironment');

var _ConfigEnvironment2 = _interopRequireDefault(_ConfigEnvironment);

var _ConfigCache = require('./ConfigCache');

var _ConfigCache2 = _interopRequireDefault(_ConfigCache);

var _ConfigPatternCache = require('./ConfigPatternCache');

var _ConfigPatternCache2 = _interopRequireDefault(_ConfigPatternCache);

var _ConfigContainer = require('./ConfigContainer');

var _ConfigContainer2 = _interopRequireDefault(_ConfigContainer);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

/**
 * @private
 * @type {ConfigContainer}
 */
const container = new _ConfigContainer2.default();

/**
 * Proxy class which automatically fills {@link Config} constructor dependencies
 * @class
 * @extends {Config}
 */
const ConfigProxy = container.proxy(_Config2.default);

/**
 * @module webpack-config
 */

exports.default = ConfigProxy;
exports.Config = ConfigProxy;

/**
 * @type {ConfigEnvironment}
 */

const environment = exports.environment = container.resolve(_ConfigEnvironment2.default);

/**
 * @type {ConfigCache}
 */
const cache = exports.cache = container.resolve(_ConfigCache2.default);

/**
 * @type {ConfigPatternCache}
 */
const patternCache = exports.patternCache = container.resolve(_ConfigPatternCache2.default);

/**
 * Returns `webpack.config.js`
 * @type {String}
 */
const FILENAME = exports.FILENAME = 'webpack.config.js';
